### ISSUE TITLE
Can't use `&mut self` in tree nodes

### ISSUE THREAD

**Issue by ckaran** [RustyYato/generic-field-projection#55]
See the comments at https://github.com/RustyYato/generic-field-projection/pull/54#discussion_r572113812_ but the gist is that as @rustyyato pointed out you can't use `&mut self` as I've been using it.  Fix this up.

**Comment by RustyYato** [comment]
I opened a related issue on rust-lang/unsafe-code-guidelines#272, because it looks like it won't be possible to provide a safe API without allocating, which is unfortunate.

**Comment by ckaran** [comment]
Yup, just saw it pop up on my feed.  I wish I could help in the decisions on this, but I don't (yet) know rust well enough to make that happen.  I'll keep reading as new posts pop up in my feed though!

**Comment by RustyYato** [comment]
The conclusion to the discussion was that nodes that hold intrusive collection nodes should be guarded to prevent `&mut _` access. One way is to do this is with a stack pin (`Pin<&T>`), something similar to `pin_utils::pin_mut` or a wrapper around `Pin<Rc<T>>`

**Comment by ckaran** [comment]
OK, I think that `Pin<Rc<T>>` might be the easiest to follow.  I think I'm going to have to make both thread-safe and thread-unsafe versions, so there will also need to be a `Pin<Arc<Mutex<T>>>`.  Right?

I'm going to admit that I'm learning as I go for some of this, the aliasing rules get weird for me...

**Comment by RustyYato** [comment]
I would postpone the thread-safe version. Rust aliasing rules mixed with thread safety is a nightmare. I think a `Mutex` would be easiest, but it basically makes a `&mut _`, so it might be unusable. I think `ReentrantMutex` (from `parking_lot`) will provide the guarantees that we want. But users of `Link` and `DoubleLink` can use those without issue.

For the tutorial, I would stick to `Pin<Rc<Node>>` if you want to show how to create a safe API, or just show the unsafe API.

**Comment by ckaran** [comment]
OK, I'll stick with `Pin<Rc<Node>>`.  Thank you!
